From eaa5f7caaa14e4b6ece18ae7714ad7659779d1e2 Mon Sep 17 00:00:00 2001
From: Pratham Malaviya <prathamm@mechasystems.com>
Date: Mon, 8 Sep 2025 11:52:50 +0530
Subject: [PATCH] drivers: extcon: Added support for vbus switch

This patch adds the support for the vbus switch gpio
in TUSB320 CC detection to control the 5V load switch
on device detection.

Signed-off-by: Pratham Malaviya <prathamm@mechasystems.com>

diff --git a/drivers/extcon/extcon-usbc-tusb320.c b/drivers/extcon/extcon-usbc-tusb320.c
index 4d08c2123e59..7de3a35ed9e5 100644
--- a/drivers/extcon/extcon-usbc-tusb320.c
+++ b/drivers/extcon/extcon-usbc-tusb320.c
@@ -82,6 +82,8 @@ struct tusb320_priv {
 	enum typec_pwr_opmode pwr_opmode;
 	struct fwnode_handle *connector_fwnode;
 	struct usb_role_switch *role_sw;
+
+	struct gpio_desc *vbus_gpiod;
 };
 
 static const char * const tusb_attached_states[] = {
@@ -259,6 +261,17 @@ static void tusb320_extcon_irq_handler(struct tusb320_priv *priv, u8 reg)
 	dev_dbg(priv->dev, "attached state: %s, polarity: %d\n",
 		tusb_attached_states[state], polarity);
 
+	/* Mecha Specific Change for Load SW */
+	if (state == 1) {
+		pr_info("TUSB LOAD SW GPIO set to HIGH\n");
+		gpiod_set_value_cansleep(priv->vbus_gpiod, 1);
+	}
+	else {
+		pr_info("TUSB LOAD SW GPIO set to LOW\n");
+		gpiod_set_value_cansleep(priv->vbus_gpiod, 0);
+	}
+	/***********************************/
+
 	extcon_set_state(priv->edev, EXTCON_USB,
 			 state == TUSB320_ATTACHED_STATE_UFP);
 	extcon_set_state(priv->edev, EXTCON_USB_HOST,
@@ -515,6 +528,8 @@ static int tusb320_probe(struct i2c_client *client)
 	const void *match_data;
 	unsigned int revision;
 	int ret;
+	u32 irq_trigger_type = IRQF_TRIGGER_FALLING;
+	struct irq_data *irq_d;
 
 	priv = devm_kzalloc(&client->dev, sizeof(*priv), GFP_KERNEL);
 	if (!priv)
@@ -527,6 +542,14 @@ static int tusb320_probe(struct i2c_client *client)
 	if (IS_ERR(priv->regmap))
 		return PTR_ERR(priv->regmap);
 
+	/* Mecha Specific Change for Load SW */
+
+	priv->vbus_gpiod = devm_gpiod_get(&client->dev,"vbus",GPIOD_OUT_LOW);
+	if (IS_ERR(priv->vbus_gpiod))
+		return PTR_ERR(priv->vbus_gpiod);
+
+	/***********************************/
+
 	ret = tusb320_check_signature(priv);
 	if (ret)
 		return ret;
@@ -568,9 +591,13 @@ static int tusb320_probe(struct i2c_client *client)
 		 */
 		tusb320_state_update_handler(priv, true);
 
+	irq_d = irq_get_irq_data(client->irq);
+	if (irq_d)
+		irq_trigger_type = irqd_get_trigger_type(irq_d);
+
 	ret = devm_request_threaded_irq(priv->dev, client->irq, NULL,
 					tusb320_irq_handler,
-					IRQF_TRIGGER_FALLING | IRQF_ONESHOT,
+					IRQF_ONESHOT | irq_trigger_type,
 					client->name, priv);
 	if (ret)
 		tusb320_typec_remove(priv);
-- 
2.25.1

